viamonvifdiscovery/service.go

```diff
@@ -0,0 +1,201 @@
+	"github.com/viam-modules/viamrtsp/viamonvif"
+	"go.viam.com/rdk/components/camera"
+	"go.viam.com/rdk/logging"
+	"go.viam.com/rdk/resource"
+	"go.viam.com/rdk/services/discovery"
+)
+
+// Model is the model for a rtsp discovery service.
+var Model = viamrtsp.Family.WithModel("discovery")
+
+var emptyCred = Creds{Username: "", Password: ""}
+
+func init() {
+	resource.RegisterService(
+		discovery.API,
+		Model,
+		resource.Registration[discovery.Service, *Config]{
+			Constructor: newDiscovery,
+		})
+}
+
+// Creds are the login credentials that a user can input.
+type Creds struct {
+	Username   string `json:"username"`
+	Password   string `json:"password"`
+	credNumber int
+}
+
+// Config is the config for the discovery service.
+type Config struct {
+	Credentials []Creds `json:"credentials"`
+}
+
+// Validate validates the discovery service.
+func (cfg *Config) Validate(_ string) ([]string, error) {
+	// check that all creds have both usernames and passwords set. Note a credential can have both fields empty
+	for _, cred := range cfg.Credentials {
+		if cred.Username != "" && cred.Password == "" {
+			return nil, fmt.Errorf("credential %v missing password", cred.Username)
+		}
+		if cred.Password != "" && cred.Username == "" {
+			return nil, fmt.Errorf("credential missing username, has password %v", cred.Password)
+		}
+	}
+	return nil, nil
+}
+
+type rtspDiscovery struct {
+	resource.Named
+	resource.AlwaysRebuild
+	Credentials []Creds
+	logger      logging.Logger
+}
+
+func newDiscovery(_ context.Context, _ resource.Dependencies,
+	conf resource.Config,
+	logger logging.Logger,
+) (discovery.Service, error) {
+	cfg, err := resource.NativeConfig[*Config](conf)
+	if err != nil {
+		return nil, err
+	}
+	dis := &rtspDiscovery{
+		Named: conf.ResourceName().AsNamed(),
+		// place the empty credentials first, so we find the insecure cameras before other cameras.
+		Credentials: append([]Creds{emptyCred}, cfg.Credentials...),
+		logger:      logger,
+	}
+
+	dis.setCredNumbers()
+	return dis, nil
+}
+
+// since camera logins can have the same username,
+// check how many repeated usernames we have and index them, to ensure cameras have unique names.
+func (dis *rtspDiscovery) setCredNumbers() {
+	// usernames do not have to be unique, so we want to add an additional label to ensure cameras have unique names.
+	for index, cred := range dis.Credentials {
+		for counter, otherCreds := range dis.Credentials {
+			if counter <= index {
+				continue
+			}
+			// increase the credNumber for later cameras.
+			if cred.Username == otherCreds.Username {
+				dis.Credentials[counter].credNumber++
+			}
+		}
+	}
+}
+
+// Close closes the discovery service.
+func (dis *rtspDiscovery) Close(_ context.Context) error {
+	return nil
+}
+
+// DiscoverResources discovers different rtsp cameras that use onvif.
+func (dis *rtspDiscovery) DiscoverResources(_ context.Context, _ map[string]any) ([]resource.Config, error) {
+	potentialCams := []resource.Config{}
```

Comment:
```suggestion
	potentialURLs := []resource.Config{}
```
or uris. A camera with one IP can have multiple stream channels associated with it.

Reply:
Calling it cameras because the function returns RTSP camera configs